### PR TITLE
Fix cast optimizer

### PR DESCRIPTION
**Pull request by raymondxyang**
* Skip the optimization when bool tensor is involved (#433 )
* Skip the optimization if the cast's output is in graph output

### CODE REVIEW

File: onnxruntime/core/optimizer/insert_cast_transformer.cc [hunk 1]
```diff
@@ -105,14 +105,25 @@ class RemoveDuplicateCastTransformer : public GraphTransformer {
...
+        auto output_args = graph.GetOutputs();
+        std::unordered_set<std::string> graph_outputs(output_args.size());
+        for (auto output_arg : output_args) {
+          graph_outputs.emplace(output_arg->Name());
+        }
         for (auto it = node.OutputNodesBegin(); it != node.OutputNodesEnd(); ++it) {
           const Node& output_node{*it};
           if (output_node.OpType() == "Cast") {
+            // Skip if the node's output is also the output of the graph
+            if (graph_outputs.find(output_node.OutputDefs()[0]->Name()) != graph_outputs.end()) {
```
Comment (weixingzhang) [inline onnxruntime/core/optimizer/insert_cast_transformer.cc]: Could the code change to check pointer of NodeArg rather than name?

Comment (pranavsharma) [inline onnxruntime/core/optimizer/insert_cast_transformer.cc]: Avoid calling output_node.OutputDefs()[0] in a loop as it doesn't depend on the loop variables?

Comment (raymondxyang) [reply onnxruntime/core/optimizer/insert_cast_transformer.cc]: isn't output_node part of the looped one?

Comment (raymondxyang) [reply onnxruntime/core/optimizer/insert_cast_transformer.cc]: Good points. actually we can break early.

Comment (pranavsharma) [reply onnxruntime/core/optimizer/insert_cast_transformer.cc]: LGTM

**Pull request merged**